bitcoin/src/bip32.rs

```diff
@@ -256,94 +350,237 @@ impl serde::Serialize for ChildNumber {
+        ///
+        /// let deriv_1 = base.extend(NormalDerivationPath::from_str("0/1").unwrap());
+        /// let deriv_2 = base.extend(&[
+        ///     NormalChildNumber::ZERO,
+        ///     NormalChildNumber::ONE
+        /// ]);
+        ///
+        /// assert_eq!(deriv_1, deriv_2);
+        /// ```
+        pub fn extend<T: AsRef<[NormalChildNumber]>>(&self, path: T) -> Self {
+            let mut new_path = self.clone();
+            new_path.0.extend_from_slice(path.as_ref());
+            new_path
+        }
+
+        /// Returns the `NormalDerivationPath` as a vector of u32 integers.
+        /// Unhardened elements are copied as is.
+        ///
+        /// ```
+        /// use bitcoin::bip32::NormalDerivationPath;
+        /// use std::str::FromStr;
+        ///
+        /// let path = NormalDerivationPath::from_str("m/84/0/0/0/1").unwrap();
+        /// const NORMAL: u32 = 0;
+        /// assert_eq!(path.to_u32_vec(), vec![84 + NORMAL, NORMAL, NORMAL, 0, 1]);
+        /// ```
+        pub fn to_u32_vec(self) -> Vec<u32> {
+            self.into_iter().map(|num| num.to_raw_number()).collect()
+        }
+
+        /// Creates a derivation path from a slice of u32s.
+        /// ```
+        /// use bitcoin::bip32::NormalDerivationPath;
+        ///
+        /// const NORMAL: u32 = 0;
+        /// let expected = vec![84 + NORMAL, NORMAL, NORMAL, 0, 1];
+        /// let path = NormalDerivationPath::from_u32_slice(expected.as_slice()).expect("Valid slice of normal numbers");
+        /// assert_eq!(path.to_u32_vec(), expected);
+        /// ```
+        pub fn from_u32_slice(numbers: &[u32]) -> Result<NormalDerivationPath, Error> {
+            numbers.iter().map(|&n| NormalChildNumber::from_index(n)).collect()
+        }
+    }
+
+    impl AsRef<[NormalChildNumber]> for NormalDerivationPath {
+        fn as_ref(&self) -> &[NormalChildNumber] { &self.0 }
+    }
+    /// Normal Derivation Path Iterator Wrapper
+    pub struct NormalIter<'a>(core::slice::Iter<'a, normal::NormalChildNumber>);
+
+    impl<'a> core::iter::Iterator for NormalIter<'a> {
+        type Item = &'a NormalChildNumber;
+        fn next(&mut self) -> Option<Self::Item> { self.0.next() }
+    }
+
+    impl<'a> core::iter::IntoIterator for &'a NormalDerivationPath {
+        type Item = &'a NormalChildNumber;
+        type IntoIter = NormalIter<'a>;
+        fn into_iter(self) -> Self::IntoIter { NormalIter(self.0.iter()) }
+    }
+    /// Normal Derivation Path IntoIterator Wrapper
+    pub struct NormalIntoIter(alloc::vec::IntoIter<NormalChildNumber>);
+
+    impl Iterator for NormalIntoIter {
+        type Item = NormalChildNumber;
+        fn next(&mut self) -> Option<Self::Item> { self.0.next() }
+    }
+
+    impl core::iter::IntoIterator for NormalDerivationPath {
+        type Item = NormalChildNumber;
+        type IntoIter = NormalIntoIter;
+        fn into_iter(self) -> Self::IntoIter { NormalIntoIter(self.0.into_iter()) }
+    }
```

Comment:
Newtypes for iterators over `NormalDerivationPath` and `DerivationPath`
@Kixunil